crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4090,8 +4090,13 @@ impl<'db> TypeInferenceBuilder<'db> {
                 // Class scopes are not visible to nested scopes, and we need to handle global
                 // scope differently (because an unbound name there falls back to builtins), so
                 // check only function-like scopes.
+                // However, there are exceptions to this rule: deferred types are visible,
+                // and type parameter scopes are evaluated lazily, thus class scopes can be visible.
```

Comment:
As discussed in other comments, I think only one of these exceptions is valid, and it's not related to lazy evaluation: in general, type parameter scopes can see names defined in their immediately-enclosing class scope.

```suggestion
                // There is one exception to this rule: type parameter scopes can see
                // names defined in an immediately-enclosing class scope.
```